### ISSUE TITLE
Installation issue: OpenMPI with LSF libdir option

### ISSUE THREAD

**Issue by robertsawko** [LLNL/spack#44255]
### Steps to reproduce the issue

@hppritcha and @naughtont3,

I am looking for guidance to fix OpenMPI package to support LSF integration. Please let me know if it's possible at all and if it is, is there anything I can do to improve on my current attempts:

## Background
My goal is to build  the following environment

```console
openfoam+scotch
gcc@13.2.0
cmake
libcatalyst+conduit
conduit~hdf5
paraview+libcatalyst+osmesa+python
openmpi+internal-pmix fabrics=auto schedulers=lsf
```
The above builds successfully when `schedulers=none`.

I wanted to have the convenience of LSF integration. So I go ahead and add LSF to `packages.yaml` (see following section), but just switching on `schedulers=lsf` breaks. When I compile OMPI manually I also need to add `--with-lsf-libdir=...` flag so I naively added this line to `config_args` in OMPI `package.py` and I can successfully build OMPI and it does appear to have LSF integration, but everything else downstream breaks.

### Error message

When compiling scotch I get the following error:
<details><summary>Error message</summary>
<pre>
1 error found in build log:
     37    -- Looking for lzma_easy_encoder in /netfs/smain01/scafellpike/local/HT05650/uxq05/shared/spack/opt/spack/linux-rhel7-broadwell/gcc-13.2.0/xz-5.4.6-jdhb7wsuhzcty64teq62u5my7z
           cxsrjl/lib/liblzma.so - found
     38    -- Looking for lzma_lzma_preset in /netfs/smain01/scafellpike/local/HT05650/uxq05/shared/spack/opt/spack/linux-rhel7-broadwell/gcc-13.2.0/xz-5.4.6-jdhb7wsuhzcty64teq62u5my7zc
           xsrjl/lib/liblzma.so
     39    -- Looking for lzma_lzma_preset in /netfs/smain01/scafellpike/local/HT05650/uxq05/shared/spack/opt/spack/linux-rhel7-broadwell/gcc-13.2.0/xz-5.4.6-jdhb7wsuhzcty64teq62u5my7zc
           xsrjl/lib/liblzma.so - found
     40    -- Found LibLZMA: /netfs/smain01/scafellpike/local/HT05650/uxq05/shared/spack/opt/spack/linux-rhel7-broadwell/gcc-13.2.0/xz-5.4.6-jdhb7wsuhzcty64teq62u5my7zcxsrjl/lib/liblzma
           .so (found version "5.4.6")
     41    -- Could NOT find MPI_C (missing: MPI_C_WORKS)
     42    -- Could NOT find MPI (missing: MPI_C_FOUND C)
  >> 43    CMake Error at src/CMakeLists.txt:126 (message):
     44      MPI required to compile PT-Scotch
     45
     46
     47    -- Configuring incomplete, errors occurred!
</pre></details>
Also, many other packages that depend on MPI break.

### Information on your system

* **Spack:** 0.23.0.dev0 (a5c8111076e13b27f955d33b1a3b6b4236b98c38)
* **Python:** 3.6.2
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo

Added LSF to `packages.yaml`:
```yaml
  lsf:
    externals:
    - spec: lsf@10.1
      prefix: "/opt/lsf/10.1"
      libdir: "/opt/lsf/10.1/linux3.10-glibc2.17-x86_64/lib"
    buildable: False
```

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/15360702/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/15360703/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by robertsawko** [comment]
Going over Spack excellent tutorials, I realised that the problem is probably that the build is being handled within an environment and Spack doesn't know that LSF installation and library can be in different places so things like `CPATH` will not be updated within the environment. Maybe a better way to fix this is to change something within LSF package?

**Comment by robertsawko** [comment]
[Summit config](https://github.com/spack/spack-configs/blob/6ef93adecac71f8ec499edd36e21bf07c25ef470/OLCF/summit/spack.yaml#L467) doesn't look too promising on this front.

**Comment by hppritcha** [comment]
I suspect the cmake find of MPI is failing because /opt/lsf/10.1/linux3.10-glibc2.17-x86_64/lib is not in the LD_LIBRARY_PATH of the shell where cmake is being run.  Kind of cludgy but you could set this env variable in your compilers.yaml.

**Comment by robertsawko** [comment]
Thanks for a response. Unfortunately, the cluster is on maintenance for the next  four days, so the first time I can check it is Friday. A lot of time to go through the tutorials over and over.

I am still not completely certain if I understand Spack environments configuration. I think whether inside and environment or not I can just
```console
spack config edit compilers
```
and add 
```yaml
compiler:
  spec: gcc@13.2.0
  ...
  environment:
    prepend_path:
      CPATH: /opt/lsf/10.1/linux3.10-glibc2.17-x86_64/include
      LD_LIBRARY_PATH: /opt/lsf/10.1/linux3.10-glibc2.17-x86_64/lib
```
I will try to get access to some other x86 LSF installation in the meantime and see if I can test it.

**Comment by hppritcha** [comment]
that's what I was thinking of trying with LD_LIBRARY_PATH.

**Comment by robertsawko** [comment]
I finally got round to doing it, but unfortunately, despite seeing added CPATH and LD_LIBRARY_PATH in Spack outputs I am getting the same problem. Basically, all downstream packages from OMPI fail: mpi4py, pt-scotch, the lot. It looks like there's no MPI compiler! I am investigating this currently.

It's really unfortunate as the `mpirun` does work as expected whereas my stack build without LSF is very clumsy to use as it needs wrappers to propagate Spack environment as described [here](https://github.com/open-mpi/ompi/issues/12556).

**Comment by hppritcha** [comment]
unfortunately i do not have access to a system with LSF installed.  @naughtont3 could you check this out on summit if its available?

**Comment by robertsawko** [comment]
I have a bit of time to investigate it today. I'd appreciate advice, but if it's too complex over github, that's understandable.

FYI information, my `package.py` added this `-with-lsf-libdir` option hard coded. I also have two environments one with `schedulers=none` and one with `schedulers=lsf`. Also, both use `fabrics=auto`. The former works as expected but it's a pain to use because no LSF integration. The latter doesn't have a working MPI. Even a test script fails, though openmpi built fine.

It seems the problem is fabrics library!
```
spack env activate envA
ldd $(spack location -i openmpi)/lib/libmpi.so | grep "libfabric"
# EMPTY
spack env activate envB
ldd $(spack location -i openmpi)/lib/libmpi.so | grep "libfabric"
libfabric.so.1 => not found
```

It looks like my `libmpi.so` has symbols from Open Fabric Interface! but `libfabrics` is not in `LD_LIBRARY_PATH` despite being inside the environment. Adding this explicitly I still have a problem with `libsharp_coll.so.4` being not found.  I don't have that in neither environment.

It seems that different fabrics support got compiled in after adding LSF. I am currently investigating whether explicit fabric in spec can help, but I believe this is still surprising behaviour, isn't it? My intuition was that these things were meant to be independent.

**Comment by robertsawko** [comment]
Yes, so being explicit about fabrics is getting me a working MPI and everything now builds, but I don't want to close it yet. I'd like to:

1. understand why explicit fabrics when `schedulers=lsf`,
1. add the `--with-lsf-libdir` option to the package.

I realise that for 2. the `packages.yaml` will need to have another variable, so I'd look into this and figure it out. Please let me know if it's worth submitting a PR for this.

**Comment by robertsawko** [comment]
Also, just for completeness for any Spacker that is in a similar position to me. I've added:

```yaml
  lsf:
    externals:
    - spec: lsf@10.1
      prefix: "/opt/lsf/10.1"
    buildable: False
  knem:
    externals:
    - spec: knem@1.1.4
      prefix: "/opt/knem-1.1.4.90mlnx3/"
    buildable: False
    version: [1.1.4]
  hcoll:
    externals:
    - spec: hcoll@4.8.3223
      prefix: "/opt/mellanox/hcoll"
    buildable: False
    version: [4.8.3223]
```
to `${SPACK_ROOT}/etc/spack/defaults/packages.yaml` and used the following spec for openmpi:
```shell
openmpi+internal-pmix fabrics=hcoll,knem,ucx schedulers=lsf
```
